api/Business/BusinessObjects/SensorBusinessObject.cs

```diff
@@ -47,11 +51,170 @@ public async Task<OperationResult> CreateSensor(Sensor record)
+                long currentUserId = long.Parse(currentUser);
+
+                if (currentUser == null)
+                {
+                    throw new Exception("Session expired");
+                }
+
+                List<Sensor> sensors = await _sensorDataAccessObject.ListSensors(currentUserId);
+                List<SensorBusinessModel> result = sensors.Select(s => new SensorBusinessModel(s)).ToList();
+
+                return result;
+            });
+        }
+
+        public async Task<OperationResult> AddData(List<SensorData> sensorData, Guid sensorUuid)
+        {
+            return await ExecuteOperation(async () =>
+            {
+                string currentUser = _httpContextAccessor.HttpContext!.User.FindFirst(ClaimTypes.NameIdentifier)!.Value;
+                long currentUserId = long.Parse(currentUser);
+
+                Sensor? sensor = await _genericDataAccessObject.GetAsync<Sensor>(sensorUuid);
+
+                if (sensor == null)
+                {
+                    throw new Exception("Sensor doesn't exist");
+                }
+                
+                long sensorId = sensor.Id;
+
+                if (currentUser == null)
+                {
+                    throw new Exception("Session expired");
+                }
+
+                if (currentUserId != sensor.UserId)
+                {
+                    throw new Exception("Unauthorized");
+                }
+
+                foreach(SensorData data in sensorData)
+                {
+                    data.SensorId= sensorId;
+                    await _genericDataAccessObject.InsertAsync(data);
+                }
+            });
+        }
+
+        public async Task<OperationResult<List<List<object>>>> ReadData(Guid sensorUuid, DateTime from, DateTime to)
+        {
+            return await ExecuteOperation(async () =>
+            {
+                string currentUser = _httpContextAccessor.HttpContext!.User.FindFirst(ClaimTypes.NameIdentifier)!.Value;
+                long currentUserId = long.Parse(currentUser);
+
+                if (currentUser == null)
+                {
+                    throw new Exception("Session expired");
+                }
+
+                Sensor? sensor = await _genericDataAccessObject.GetAsync<Sensor>(sensorUuid);
+          
+                if (sensor == null)
+                {
+                    throw new Exception("sensor does not exist");
+                }
+
+                if (!sensor.IsPublic && currentUserId != sensor.UserId)
+                {
+                    throw new Exception("Sensor is not accessible");
+                }
+
+                List<SensorData> result = await _sensorDataAccessObject.ReadData(currentUserId, sensorUuid, from, to);
+                List<List<object>> sensorData = new List<List<object>>();
+
+                foreach (SensorData data in result)
```

Comment:
can't you use a select on the result to map the data to your model here instead of a foreach?

Reply:
i can. replaced it now